### ISSUE TITLE
Driver error

### ISSUE THREAD

**Issue by hm-sarah** [StanleyInnovation/segway_v3#13]
The onboard PC can't communicate with the RMP 220 V3 anymore. There is a driver error/missing preventing status updates or communication with RMP 220 V3. I had the navigation demo pre-installed. How can I retrieve/recover lost drivers?

**Comment by PatrickHussey** [comment]
What is the error that is being thrown? You can get it from the ROS logfile or from opening a terminal and running:

1. segstop
2. segstart
3. continuously run the segchk command until you see the error print out 

The driver is part of the segway_ros package in the segway_v3 repo. So there shouldn't be a way to lose it. First try to ssh (or connect a monitor and keyboard) to the onboard PC and ping the embedded system (ping 10.66.171.5). If it responds then the ethernet connection is fine, if not check that the ethernet cable from the UI harness is plugged into eth0-eth4 (but not eth5 it is not part of the bridge and is on the front closest to the USB connections). 

If you still can't ping it then try running ifconfig (you should see br0 is 10.66.171.4). If all that looks good then a configuration or something might have been changed from what we delivered. Here is the configuration that should be in segway_config.sh on the onboard PC and on the remote PC in the package segway_v3/segway_v3_config:

`# This configures the environment variables for a segway rmp V3 simulation
# This is necessary to run before starting the simulation 
#

# If there is an onboard PC powered by the system this will run the watchdog
# to make sure it gets gracefully shutdown before the system power cuts out.
export SEGWAY_POWERS_PC_ONBOARD=true

#Default Segway RMP_V3 network 
export SEGWAY_IP_ADDRESS=10.66.171.5
export SEGWAY_IP_PORT_NUM=8080

# The reference frame for all the accessories and sensors
export SEGWAY_PARENT_LINK=base_chassis_link

#Determines if move base should run in balance mode, if false runs in tractor
export SEGWAY_RUNS_IN_BALANCE_MODE=false
# Determines if Segway platform is equipped with a BSA almost all are expect the RMP210
export SEGWAY_HAS_BSA=true

# Platform name for specific URDF models and configurations; the base platform is
# used for drivers, sim controllers, etc; the platform name is for special URDF/Sensor
# configurations. In general they should be the same unless we built you a custom system
export SEGWAY_BASE_PLATFORM=RMP_220
export SEGWAY_PLATFORM_NAME=RMP_220
export SEGWAY_IS_OMNI=false

# Defines the location of the sensors frames and the UI box
# they all move together for the navigator elite. This is the point at the
# base of the sensor pedestal aligned with the base_link
export NAVIGATOR_ELITE_XYZ="-0.1801 0 0.060625"

#Only the RMP220 and RMP210 can have caster plates so do not enable them if using another model
export SEGWAY_HAS_REAR_CASTER=true
export SEGWAY_HAS_FRONT_CASTER=true

#Determines if the platform should use 2D or 3D odometry
export SEGWAY_USE_2D_ODOMETRY=true
# Determines if we should use robot_localization EKF or the onboard odometry;
# set to true for onboard odometry
export SEGWAY_USE_PLATFORM_ODOMETRY=true
# determines if GPS is used for local or global odometry (requires a GPS data source)
export SEGWAY_USE_GPS_FOR_LOCAL_ODOM=false
export SEGWAY_USE_GPS_FOR_GLOBAL_LOCALIZATION=false

# Joystick configurations for onboard joystick
# Joystick configurations for onboard joystick
export SEGWAY_HAS_ONBOARD_JOY=false
export SEGWAY_JOY_DEV=/dev/input/js0
export SEGWAY_JOY_MAPPING=xbox360
export SEGWAY_JOY_DEADZONE=0.1

# Sets whether the costmaps use 3D voxel grid layer (need a 3D sensor enabled)
export SEGWAY_USE_VOXEL_GRID=false

# External IMU configuration;
export SEGWAY_HAS_EXT_IMU=false
export SEGWAY_HAS_UM7_IMU=false
export SEGWAY_UM7_DEV="/dev/serial/by-path/pci-0000:00:14.0-usb-0:3:1.0-port0"
export EXT_IMU_XYZ="-0.25203 0.130 0.0715"
export EXT_IMU_RPY="3.1415 0 3.1415"

# GPS configuration
export SEGWAY_HAS_GPS=false
export SEGWAY_HAS_NEO7=false
export SEGWAY_NEO7_DEV="/dev/serial/by-path/pci-0000:00:14.0-usb-0:4:1.0-port0"
export GPS_XYZ="0.5 0 0.25409"
export GPS_RPY="0 0 0"

# 2D scanner configuration for launch files; should change hokuyo
# to sick if using a SICK LMS1XX
export SEGWAY_HAS_ONE_2D_LASER=false
export SEGWAY_HAS_SECOND_2D_LASER=false

# Hokuyo configuration (only supports two by default) watch 
# wavelength on multi-laser systems
export SEGWAY_LASER1_IS_HOKUYO=false
export SEGWAY_LASER1_IS_SICK_TIM=false
export SEGWAY_LASER1_IP=10.66.171.8
export SEGWAY_LASER1_PORT=2112
#Uncomment below and comment line above for hokuyo
#export SEGWAY_LASER1_PORT=10940 
export LASER1_XYZ="0.62204 0 0.25409"
export LASER1_RPY="0 0 0"
export LASER1_MAX_RANGE=20.0
export LASER1_MIN_RANGE=0.01
export LASER1_PREFIX="front"

export SEGWAY_LASER2_IS_HOKUYO=false
export SEGWAY_LASER2_IS_SICK_TIM=false
export SEGWAY_LASER2_IP=10.66.171.9
export SEGWAY_LASER2_PORT=2112
#Uncomment below and comment line above for hokuyo
#export SEGWAY_LASER1_PORT=10940 
export LASER2_XYZ="-0.62204 0 0.25409"
export LASER2_RPY="0 0 3.1415"
export LASER2_MAX_RANGE=20.0
export LASER2_MIN_RANGE=0.01
export LASER2_PREFIX="rear"

# Segway has KINECT
export SEGWAY_HAS_KINECT=false
export SEGWAY_HAS_KINECT_360=false
#export SEGWAY_HAS_KINECT_ONE=false not supported unless custom platform
export RGBD_SENSOR_XYZ="0 0 1.0"
export RGBD_SENSOR_RPY="0 0 0"

# Segway has VLP16
export SEGWAY_HAS_VLP16=false
export SEGWAY_VLP16_IP=192.168.1.201
export SEGWAY_VLP16_PORT=2368
export VLP16_XYZ="0.0254 0 0.42760744"
export VLP16_RPY="0 0 0"

# Machine vision
export SEGWAY_HAS_FLEA3=false
export SEGWAY_FLEA3_SN='14472220'
export SEGWAY_FLEA3_IS_CALIBRATED=false
export FLEA3_XYZ="0.0518 0 0.35508750"
export FLEA3_RPY="0 0 0"`

**Comment by hm-sarah** [comment]
![si](https://cloud.githubusercontent.com/assets/7173400/19365673/e15e2fae-91a4-11e6-8904-9953408bb501.png)
![si2](https://cloud.githubusercontent.com/assets/7173400/19365672/e15adad4-91a4-11e6-8d3f-4d0d460868d6.png)

These are the errors that I am receiving, when I send this command: roslaunch segway_bringup stanley_innovation_system.launch

**Comment by PatrickHussey** [comment]
The computer was delivered with the upstart job installed (it runs when the platform is booted up and starts all the ROS stuff).  

It looks like you are trying to launch again, did you stop the service first?

open a terminal and run segstop
then try launching if you want/need to do it manually

**Comment by hm-sarah** [comment]
Hi Patrick, 

I've discovered a node crash upon startup- Perhaps this is the culprit (any idea what it does?)
![si5](https://cloud.githubusercontent.com/assets/7173400/19366348/e89394aa-91a7-11e6-9f41-306cc401ec03.png)
![si6](https://cloud.githubusercontent.com/assets/7173400/19366349/e8998e14-91a7-11e6-8336-372cbe3da93e.png)

**Comment by PatrickHussey** [comment]
It is the velodyne nodelet manager crashing. Unfortunately we did not integrate the Velodyne for you so I can't really say what is wrong, maybe the network configuration for the velodyne?

If you set line line in segway_config.sh
export SEGWAY_HAS_VLP16=false

Re-source the workspace and re-launch does everything work as expected? 

You can always launch the Velodyne drivers separately to debug. The one we have is for a particular build of code for the VLP-16 which we have in inventory and ship with the navigator elite sensor packages. The standard drivers; which we don't support are here:

https://github.com/ros-drivers/velodyne

**Comment by hm-sarah** [comment]
We still have the error, the Velodyne was already false

![si2](https://cloud.githubusercontent.com/assets/7173400/19367111/df73649c-91aa-11e6-92b6-b48449cc320d.png)

**Comment by PatrickHussey** [comment]
Both the error logs you posted indicate that the Velodyne driver is the culprit and is somehow being launched. Try closing the terminal, opening a new one and re-launching. The environment variables persist in a shell while it remains open.

When you open the new one and source the workspace run:
echo $SEGWAY_HAS_VLP16

it should return false

**Comment by PatrickHussey** [comment]
Also try:
close all terminals and open a new one. Run:
rostopic list

It should not return anything. If it does then the service is already running and you need to stop it.

**Comment by hm-sarah** [comment]
![si7](https://cloud.githubusercontent.com/assets/7173400/19367269/73e8142e-91ab-11e6-932d-25d678f0c2de.png)

The Velodyne has returned false. 
This is our network configuration, Is there anything out of the ordinary?

**Comment by PatrickHussey** [comment]
No looks right, can you ping the platform at 10.66.171.5?

and what do you get returned when you run:
echo $ROBOT_NETWORK
(should be br0)
echo $ROS_IP
(should be 10.66.171.4)
echo $ROS_MASTER_URI
(should be http://10.66.171.4:11311/)

Still have no idea why it would be complaining about the velodyne nodelet manager? Have you modified the source code on the on-board PC that was delivered in any way? I believe you indicated that you were able to drive it around in a previous post with an onboard joystick, and it was definitely tested before it left (I did the testing). 

I am honestly at a loss as to why it doesn't work now, unless something was changed that I am unaware of.

**Comment by hm-sarah** [comment]
I can ping the platform, but when I run 
echo $ROBOT_NETWORK
echo $ROS_IP
echo $ROS_MASTER_URI
all return nothing

**Comment by PatrickHussey** [comment]
You need to source the workspace for the environment hooks to get sourced.

**Comment by PatrickHussey** [comment]
**cd segway_ws**

and then

**source ./devel/setup.bash**

or use the bash alias defined in ~/.bashrc

**sws**

then try and echo those and launch the drivers

**Comment by hm-sarah** [comment]
We already sourced the workspace, all return nothing, but I am able to ping 10.66.171.5 ...

**Comment by PatrickHussey** [comment]
Is there a segway_v3_network package in the source directory?

doesn't sound like it.

if not run these commands in a new terminal:

```
cd ~/segway_ws/src
git clone https://github.com/StanleyInnovation/segway_v3_network.git
gedit segway_v3_network/env-hooks/50.segway_network_config.sh
```

in gedit:
change **eth0** to **br0**
save and close the file

back in the terminal
`cd ../ && catkin_make`

close the terminal and open a new one
```
cd ~/segway_ws
sws
```

then try and echo those and launch the drivers

thats probably the issue, but still does not explain the velodyne nodelet

**Comment by PatrickHussey** [comment]
did this get resolved?

**Comment by cangjiaxaun** [comment]
Hi, Patrick. I also meet communication error. Our platform is a dual arm platform.
First we tried to launch the program without the arms.
I rename the `/usr/sbin/segway_core_start`. So I can run this file in the terminal. I also tried `roslaunch segway_bringup stanley_innovation_system.launch` without run the "`segway_core_start`" program. 
Here is the screen shot when I run the program in the terminal:
![error1](https://cloud.githubusercontent.com/assets/15178808/19391877/05ff5a2e-9261-11e6-94d9-f93f2ce3cffb.png)
![error2](https://cloud.githubusercontent.com/assets/15178808/19391885/0abad750-9261-11e6-9e0b-65eb28795762.png)
![error3](https://cloud.githubusercontent.com/assets/15178808/19391888/0c2bcf9a-9261-11e6-9797-78dafac7071c.png)
**I think we have the third error because the first one, that is the communication problem.**
And I got a new warning message as shown in the following pic, all the other error messages are the same.
![img-20161014-wa0029](https://cloud.githubusercontent.com/assets/15178808/19392028/962824a0-9261-11e6-9043-f3a96c73774b.jpg)

Plus, we tried to run the program with the arm connected
And I got the same warning message as above, as shown in the following picture, and all other error messages are the same 
![error4](https://cloud.githubusercontent.com/assets/15178808/19392499/9602a688-9263-11e6-9202-965ae68a74e1.png)
-----------------------------------------------------------------------------------------------------------------I I tried to ping the IP address 10.66.171.5 and 10.66.171.4, both are success. And all the echo return the right result. Would you please help me with my problem? Thanks in advance!

**Comment by PatrickHussey** [comment]
Yours is a different issue of incompatible embedded firmware and ROS driver version.

I will create a separate thread

**Comment by cangjiaxaun** [comment]
OK, see you in that thread:)

**Comment by cangjiaxaun** [comment]
I opened a new issue
